src/AST/Translator.Entity.dfy

```diff
@@ -0,0 +1,298 @@
+    reads *
+    decreases ASTHeight(attrs)
+  {
+    if attrs == null then
+      Success([])
+    else
+      var name := TypeConv.AsString(attrs.Name);
+      var args :- Seq.MapResult(ListUtils.ToSeq(attrs.Args), Expr.TranslateExpression);
+      assume ASTHeight(attrs.Prev) < ASTHeight(attrs);
+      var rest :- TranslateAttributes(attrs.Prev);
+      Success([E.Attribute.Attribute(TranslateAttributeName(name), args)] + rest)
+  }
+
+  function TranslateMemberEntityInfo(md: C.MemberDecl): (e: TranslationResult<E.EntityInfo>)
+    reads *
+  {
+    var name :- TranslateName(md.FullName);
+    var attrs :- TranslateAttributes(md.Attributes);
+    var loc := TranslateLocation(md.tok);
+    Success(E.EntityInfo(name, location := loc, attrs := attrs, members := []))
+  }
+
+  function TranslateField(f: C.Field): (d: TranslationResult<E.Entity>)
+    reads *
+  {
+    var kind := if f.IsMutable then E.Var else E.Const;
+    var ei :- TranslateMemberEntityInfo(f);
+    if f is C.ConstantField then
+      var init :- Expr.TranslateExpression((f as C.ConstantField).Rhs);
+      Success(E.Definition(ei, E.Definition.Field(E.Field.Field(kind, Some(init)))))
+    else
+      Success(E.Definition(ei, E.Definition.Field(E.Field.Field(kind, None))))
+  }
+
+  function TranslateMethod(m: C.Method): (d: TranslationResult<E.Entity>)
+    reads *
+  {
+    var body :- Expr.TranslateOptionalStatement(m.Body);
+    var req :- Seq.MapResult(ListUtils.ToSeq(m.Req), (ae: C.AttributedExpression) reads * => Expr.TranslateExpression(ae.E));
+    var ens :- Seq.MapResult(ListUtils.ToSeq(m.Ens), (ae: C.AttributedExpression) reads * => Expr.TranslateExpression(ae.E));
+    var def := if m is C.Constructor then
+                 E.Constructor(req := req, ens := ens, body := body)
+               else
+                 E.Method(req := req, ens := ens, body := body);
+    var ei :- TranslateMemberEntityInfo(m);
+    Success(E.Definition(ei, E.Callable(def)))
+  }
+
+  function TranslateFunction(f: C.Function): (d: TranslationResult<E.Entity>)
+    reads *
+  {
+    var body :- Expr.TranslateOptionalExpression(f.Body);
+    var req :- Seq.MapResult(ListUtils.ToSeq(f.Req), (ae: C.AttributedExpression) reads * => Expr.TranslateExpression(ae.E));
+    var ens :- Seq.MapResult(ListUtils.ToSeq(f.Ens), (ae: C.AttributedExpression) reads * => Expr.TranslateExpression(ae.E));
+    var ei :- TranslateMemberEntityInfo(f);
+    Success(E.Definition(ei, E.Callable(E.Function(req := req, ens := ens, body := body))))
+  }
+
+  function TranslateMemberDecl(md: C.MemberDecl): (d: TranslationResult<E.Entity>)
+    reads *
+  {
+    if md is C.Field then
+      TranslateField(md)
+    else if md is C.Function then
+      TranslateFunction(md)
+    else if md is C.Method then
+      TranslateMethod(md)
+    else
+      Failure(Invalid("Unsupported member declaration type: " + TypeConv.AsString(md.FullName)))
```

Comment:
We should add an unsupported entity constructor and return `Success` with it here.